presto-kafka/src/main/java/io/prestosql/plugin/kafka/encoder/RowEncoder.java

```diff
@@ -0,0 +1,185 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.prestosql.plugin.kafka.encoder;
+
+import com.google.common.primitives.Shorts;
+import com.google.common.primitives.SignedBytes;
+import io.prestosql.spi.block.Block;
+import io.prestosql.spi.connector.ConnectorSession;
+import io.prestosql.spi.type.SqlDate;
+import io.prestosql.spi.type.SqlTime;
+import io.prestosql.spi.type.SqlTimeWithTimeZone;
+import io.prestosql.spi.type.SqlTimestamp;
+import io.prestosql.spi.type.SqlTimestampWithTimeZone;
+import io.prestosql.spi.type.TimestampType;
+import io.prestosql.spi.type.TimestampWithTimeZoneType;
+import io.prestosql.spi.type.Type;
+
+import java.nio.ByteBuffer;
+
+import static io.prestosql.spi.type.BigintType.BIGINT;
+import static io.prestosql.spi.type.BooleanType.BOOLEAN;
+import static io.prestosql.spi.type.DateType.DATE;
+import static io.prestosql.spi.type.DoubleType.DOUBLE;
+import static io.prestosql.spi.type.IntegerType.INTEGER;
+import static io.prestosql.spi.type.RealType.REAL;
+import static io.prestosql.spi.type.SmallintType.SMALLINT;
+import static io.prestosql.spi.type.TimeType.TIME;
+import static io.prestosql.spi.type.TimeWithTimeZoneType.TIME_WITH_TIME_ZONE;
+import static io.prestosql.spi.type.TinyintType.TINYINT;
+import static io.prestosql.spi.type.VarbinaryType.isVarbinaryType;
+import static io.prestosql.spi.type.Varchars.isVarcharType;
+import static java.lang.Float.intBitsToFloat;
+import static java.lang.Math.toIntExact;
+import static java.lang.String.format;
+
+public interface RowEncoder
+{
+    byte[] toByteArray();
+
+    void clear();
+
+    default RowEncoder putValueFromBlock(EncoderColumnHandle columnHandle, ConnectorSession session, Block block, int position)
```

Comment:
The `position` argument here represents the index of the column from the table schema, which is not necessarily the same as the position in the result set, so we will still need a mapping (though it would be just an array lookup).

> Furthermore, we can guarantee that columns are being provided in order.
This would unlock certain optimizations in encoders like CSV or RAW, which
would be writing to a byte buffer directly.

I don't know if that's necessarily the case.  It would be pretty constraining for the API to require values be provided in the order columns appear in the table schema.

Perhaps we would be better off passing the entire page, instead of individual blocks, and let the encoder decide how to iterate over columns.

Reply:
> The `position` argument here represents the index of the column from the table schema

block+position represents a value (position is the row)

Reply:
> `void appendValue(Block block, int position)`

As discussed offline w @charlesjmorgan  -- `appendNextColumnValue`

> I don't know if that's necessarily the case. It would be pretty constraining for the API to require values be provided in the order columns appear in the table schema.

Why is that constraining? 
Wha would be the advantage of relaxing this?

Reply:
> > The `position` argument here represents the index of the column from the table schema
> 
> block+position represents a value (position is the row)

Sorry, that was a dumb comment.

Where my head was at is that you might have a table schema like `foo (x varchar, y bigint, z boolean)` and you run `INSERT INTO foo (z, y) VALUES (true, 42)`, so if you don't specify the column, you need to perform a mapping for the column index from the result set schema to the table schema.

> Why is that constraining?

In the example above, you'd have to append `null`, then `42`, then `true`.  It seems like it requires a lot of work for the API client.

Reply:
> this would allow encoder to precalculate ("compile") encoding strategies for each
of the columns

By this do you (@findepi) mean that the encoder could create a list of write functions for the different column types similar to the jdbc connector?

Are columns in pages always in the same order?

Reply:
I'm worried that key columns and message columns being in the same page might mess up the ordering

Reply:
It looks like pages always have the columns in the same (correct) order which is good. I could track the index of the key and message columns separately with two different ints.

>It seems like it requires a lot of work for the API client.

@aalbu I think that this is a good point. My only worry is that if we pass the entire page to the key/message encoder then the page includes both the key and message columns so it would need some way to distinguish these. Maybe we could pass a range of index values in the method too? Or would it be better to construct two `Block[]` for the key and message columns and then pass each of those once to their respective encoders to minimize the number of api calls needed.

Reply:
also I'm not sure if the name `append...` really fits if we specify the position to put the value at

Reply:
> It looks like pages always have the columns in the same (correct) order which is good

Ah, my incorrect assumption was that only the columns specified in the insert are going to be present on the page.  That makes my concern irrelevant.

Reply:
> By this do you (@findepi) mean that the encoder could create a list of write functions for the different column types similar to the jdbc connector?

it can, if it wishes to do so

i am not saying "do it". I am just saying "it is possible"

> Are columns in pages always in the same order?

yes

Reply:
> @aalbu I think that this is a good point. My only worry is that if we pass the entire page to the key/message encoder then the page includes both the key and message columns so it would need some way to distinguish these.

That's a very good point, which actually rules out passing full `Page` (+ position)